File: manual-verification-required-never-cleared/node/clear-manual-verification.js
/**
 * Surface and clear BigCommerce orders stuck on Manual Verification Required.
 *
 * A fraud-screening app (FraudLabs Pro, NoFraud, Signifyd, Kount) or an ERP
 * connector writes status_id 12 to an order when it flags a REVIEW verdict.
 * The human review then happens inside that app's own dashboard, so nothing
 * tells BigCommerce the order was approved. This never auto-transitions an
 * order on elapsed time. It only reports orders that already carry an
 * explicit human-approval marker in staff_notes or messages, with a
 * non-declined transaction, and only writes when DRY_RUN=false.
 * Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/manual-verification-required-never-cleared/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example-store";
const TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "dummy-token";
const BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/`;
const MIN_DATE_MODIFIED = process.env.MIN_DATE_MODIFIED || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const MANUAL_VERIFICATION_STATUS_ID = 12;
const AWAITING_FULFILLMENT_STATUS_ID = 11;

const APPROVAL_RE = /\b(approved|cleared|verified)\b/i;
const DECLINED_TXN_STATUSES = new Set(["declined", "void"]);
const OK_TXN_STATUSES = new Set([null, undefined, "approved", "captured"]);

function parseDate(value) {
  if (!value) return null;
  const d = new Date(value);
  return Number.isNaN(d.getTime()) ? null : d;
}

/**
 * Pure decision logic, no I/O. Returns "clear", "hold", or "skip".
 *
 * - "skip": the order is not on status_id 12, nothing to do here.
 * - "hold": still needs a human, or the transaction looks unsafe to clear.
 * - "clear": an explicit human-approval marker was found in staff_notes or a
 *   message timestamped after date_modified, and the transaction is not
 *   declined or voided.
 */
export function decideClearable(order, messages, transactionStatus) {
  if (order.status_id !== MANUAL_VERIFICATION_STATUS_ID) return "skip";
  if (DECLINED_TXN_STATUSES.has(transactionStatus)) return "hold";

  const modifiedAt = parseDate(order.date_modified);
  let hasMarker = false;

  const note = order.staff_notes || "";
  if (APPROVAL_RE.test(note)) hasMarker = true;

  if (!hasMarker) {
    for (const msg of messages || []) {
      const text = msg.message || "";
      if (!APPROVAL_RE.test(text)) continue;
      const createdAt = parseDate(msg.date_created);
      if (modifiedAt === null || createdAt === null || createdAt >= modifiedAt) {
        hasMarker = true;
        break;
      }
    }
  }

  if (hasMarker && OK_TXN_STATUSES.has(transactionStatus)) return "clear";
  return "hold";
}

async function bc(method, path, body) {
  const res = await fetch(BASE + path.replace(/^\//, ""), {
    method,
    headers: { "X-Auth-Token": TOKEN, "Content-Type": "application/json", "Accept": "application/json" },
    body: body ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const text = await res.text();
  return text ? JSON.parse(text) : null;
}

async function* ordersPendingVerification() {
  let page = 1;
  while (true) {
    const qs = new URLSearchParams({ status_id: String(MANUAL_VERIFICATION_STATUS_ID), limit: "250", page: String(page) });
    if (MIN_DATE_MODIFIED) qs.set("min_date_modified", MIN_DATE_MODIFIED);
    const batch = (await bc("GET", `/v2/orders?${qs.toString()}`)) || [];
    if (batch.length === 0) return;
    for (const order of batch) yield order;
    if (batch.length < 250) return;
    page += 1;
  }
}

async function orderMessages(orderId) {
  return (await bc("GET", `/v2/orders/${orderId}/messages`)) || [];
}

async function transactionStatusFor(orderId) {
  const txns = (await bc("GET", `/v2/orders/${orderId}/transactions`)) || [];
  for (const t of txns) {
    const status = (t.status || "").toLowerCase();
    if (DECLINED_TXN_STATUSES.has(status)) return status;
  }
  for (const t of txns) {
    const status = (t.status || "").toLowerCase();
    if (status === "approved" || status === "captured") return status;
  }
  return null;
}

async function clearOrder(orderId) {
  await bc("PUT", `/v2/orders/${orderId}`, { status_id: AWAITING_FULFILLMENT_STATUS_ID });
  await bc("POST", `/v2/orders/${orderId}/messages`, {
    message: "Cleared Manual Verification Required to Awaiting Fulfillment "
      + "after finding an explicit staff approval marker and a non-declined transaction.",
    status_id: AWAITING_FULFILLMENT_STATUS_ID,
  });
}

export async function run() {
  let cleared = 0;
  let held = 0;
  for await (const order of ordersPendingVerification()) {
    const orderId = order.id;
    const fullOrder = (await bc("GET", `/v2/orders/${orderId}`)) || order;
    const messages = await orderMessages(orderId);
    const txnStatus = await transactionStatusFor(orderId);

    const decision = decideClearable(fullOrder, messages, txnStatus);
    if (decision === "skip") continue;
    if (decision === "hold") { held++; continue; }

    console.log(`Order ${orderId} clearable. ${DRY_RUN ? "would clear" : "clearing"}`);
    if (!DRY_RUN) await clearOrder(orderId);
    cleared++;
  }
  console.log(`Done. ${cleared} order(s) ${DRY_RUN ? "to clear" : "cleared"}, ${held} held for review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: manual-verification-required-never-cleared/python/clear_manual_verification.py
"""Surface and clear BigCommerce orders stuck on Manual Verification Required.

A fraud-screening app (FraudLabs Pro, NoFraud, Signifyd, Kount) or an ERP
connector writes status_id 12 to an order when it flags a REVIEW verdict.
The human review then happens inside that app's own dashboard, so nothing
tells BigCommerce the order was approved. This never auto-transitions an
order on elapsed time. It only reports orders that already carry an explicit
human-approval marker in staff_notes or messages, with a non-declined
transaction, and only writes when DRY_RUN=false. Safe to run again and again.

Guide: https://www.allanninal.dev/bigcommerce/manual-verification-required-never-cleared/
"""
import os
import re
import logging
import requests
from datetime import datetime
from email.utils import parsedate_to_datetime

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("clear_manual_verification")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/"
MIN_DATE_MODIFIED = os.environ.get("MIN_DATE_MODIFIED")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

MANUAL_VERIFICATION_STATUS_ID = 12
AWAITING_FULFILLMENT_STATUS_ID = 11

APPROVAL_RE = re.compile(r"\b(approved|cleared|verified)\b", re.IGNORECASE)
DECLINED_TXN_STATUSES = {"declined", "void"}
OK_TXN_STATUSES = {None, "approved", "captured"}


def bc(method, path, **kwargs):
    r = requests.request(
        method, BASE + path.lstrip("/"),
        headers={"X-Auth-Token": TOKEN, "Content-Type": "application/json", "Accept": "application/json"},
        timeout=30, **kwargs,
    )
    r.raise_for_status()
    return r.json() if r.content else None


def _parse_date(value):
    """Best-effort parse of an ISO-8601 or RFC-2822 timestamp. Returns None on failure."""
    if not value:
        return None
    try:
        return parsedate_to_datetime(value)
    except (TypeError, ValueError):
        try:
            return datetime.fromisoformat(value)
        except ValueError:
            return None


def decide_clearable(order, messages, transaction_status):
    """Pure decision logic, no I/O. Returns 'clear', 'hold', or 'skip'.

    - 'skip': the order is not on status_id 12, nothing to do here.
    - 'hold': still needs a human, or the transaction looks unsafe to clear.
    - 'clear': an explicit human-approval marker was found in staff_notes or a
      message timestamped after date_modified, and the transaction is not
      declined or voided.
    """
    if order.get("status_id") != MANUAL_VERIFICATION_STATUS_ID:
        return "skip"
    if transaction_status in DECLINED_TXN_STATUSES:
        return "hold"

    modified_at = _parse_date(order.get("date_modified"))
    has_marker = False

    note = order.get("staff_notes") or ""
    if APPROVAL_RE.search(note):
        has_marker = True

    if not has_marker:
        for msg in messages or []:
            text = msg.get("message") or ""
            if not APPROVAL_RE.search(text):
                continue
            created_at = _parse_date(msg.get("date_created"))
            if modified_at is None or created_at is None or created_at >= modified_at:
                has_marker = True
                break

    if has_marker and transaction_status in OK_TXN_STATUSES:
        return "clear"
    return "hold"


def orders_pending_verification():
    """Yield every order currently on status_id 12, paginating with page/limit."""
    page = 1
    while True:
        params = {"status_id": MANUAL_VERIFICATION_STATUS_ID, "limit": 250, "page": page}
        if MIN_DATE_MODIFIED:
            params["min_date_modified"] = MIN_DATE_MODIFIED
        batch = bc("GET", "/v2/orders", params=params) or []
        if not batch:
            return
        for order in batch:
            yield order
        if len(batch) < 250:
            return
        page += 1


def order_messages(order_id):
    return bc("GET", f"/v2/orders/{order_id}/messages") or []


def transaction_status_for(order_id):
    """Reduce an order's transactions to a single status string, or None.

    A declined or voided transaction always wins so it can never be masked by
    an approval marker. Otherwise prefer an approved/captured transaction.
    """
    txns = bc("GET", f"/v2/orders/{order_id}/transactions") or []
    for t in txns:
        status = (t.get("status") or "").lower()
        if status in DECLINED_TXN_STATUSES:
            return status
    for t in txns:
        status = (t.get("status") or "").lower()
        if status in {"approved", "captured"}:
            return status
    return None


def clear_order(order_id):
    """Move the order to Awaiting Fulfillment and leave an audit message."""
    bc("PUT", f"/v2/orders/{order_id}", json={"status_id": AWAITING_FULFILLMENT_STATUS_ID})
    bc("POST", f"/v2/orders/{order_id}/messages", json={
        "message": "Cleared Manual Verification Required to Awaiting Fulfillment "
                   "after finding an explicit staff approval marker and a non-declined transaction.",
        "status_id": AWAITING_FULFILLMENT_STATUS_ID,
    })


def run():
    cleared = 0
    held = 0
    for order in orders_pending_verification():
        order_id = order["id"]
        full_order = bc("GET", f"/v2/orders/{order_id}") or order
        messages = order_messages(order_id)
        txn_status = transaction_status_for(order_id)

        decision = decide_clearable(full_order, messages, txn_status)
        if decision == "skip":
            continue
        if decision == "hold":
            held += 1
            continue

        log.info("Order %s clearable. %s", order_id, "would clear" if DRY_RUN else "clearing")
        if not DRY_RUN:
            clear_order(order_id)
        cleared += 1

    log.info("Done. %d order(s) %s, %d held for review.",
              cleared, "to clear" if DRY_RUN else "cleared", held)


if __name__ == "__main__":
    run()
